ctapipe/tools/train_energy_regressor.py

```diff
@@ -0,0 +1,156 @@
+
+    flags = {
+        **flag(
+            "overwrite",
+            "TrainEnergyRegressor.overwrite",
+            "Overwrite output existing output files",
+            "Don't overwrite existing output files",
+        ),
+    }
+
+    aliases = {
+        ("i", "input"): "TableLoader.input_url",
+        ("o", "output"): "TrainEnergyRegressor.output_path",
+        "cv-output": "CrossValidator.output_path",
+    }
+
+    classes = [
+        TableLoader,
+        EnergyRegressor,
+        CrossValidator,
+    ]
+
+    def setup(self):
+        """
+        Initialize components from config
+        """
+        self.loader = TableLoader(
+            parent=self,
+            load_dl1_images=False,
+            load_dl1_parameters=True,
+            load_dl2=True,
+            load_simulated=True,
+            load_instrument=True,
+        )
+
+        self.regressor = EnergyRegressor(self.loader.subarray, parent=self)
+        self.cross_validate = CrossValidator(
+            parent=self, model_component=self.regressor
+        )
+        self.rng = np.random.default_rng(self.random_seed)
+
+    def start(self):
+        """
+        Train models per telescope type.
+        """
+
+        types = self.loader.subarray.telescope_types
+        self.log.info("Inputfile: %s", self.loader.input_url)
+        self.log.info("Training models for %d types", len(types))
+        for tel_type in types:
+            self.log.info("Loading events for %s", tel_type)
+            table = self._read_table(tel_type)
+
+            self.log.info("Train on %s events", len(table))
+            self.cross_validate(tel_type, table)
+
+            self.log.info("Performing final fit for %s", tel_type)
+            self.regressor.fit(tel_type, table)
+            self.log.info("done")
+
+    def _read_table(self, telescope_type):
+        table = self.loader.read_telescope_events([telescope_type])
+
+        self.log.info("Events read from input: %d", len(table))
+        mask = self.regressor.quality_query.get_table_mask(table)
+        table = table[mask]
+        self.log.info("Events after applying quality query: %d", len(table))
+
+        table = self.regressor.feature_generator(table)
+
+        feature_names = self.regressor.features + [self.regressor.target]
+        table = table[feature_names]
+
+        valid = check_valid_rows(table)
+        if np.any(~valid):
+            self.log.warning("Dropping non-predictable events.")
+            table = table[valid]
+
+        if self.n_events is not None:
+            n_events = min(self.n_events, len(table))
+            idx = self.rng.choice(len(table), n_events, replace=False)
+            idx.sort()
+            table = table[idx]
+
+        return table
+
+    def finish(self):
+        """
+        Write-out trained models and cross-validation results.
+        """
+        self.log.info("Writing output")
+        self.regressor.write(self.output_path, overwrite=self.overwrite)
```

Comment:
We should move this to somewhere earlier. I just waited for the training to finish, only to then fail here.

Reply:
But we can think about that in #2111, because all Tools (might) have this issue.